infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutAssetDao.kt

```diff
@@ -280,6 +281,26 @@ abstract class LayoutAssetDao<T : LayoutAsset<T>>(
         return jdbcTemplate.queryOptional(officialVersionAtMomentSql, params, ::toRowVersion)
     }
 
+    override fun fetchOfficialRowVersionForPublishingInBranch(
+        branch: LayoutBranch,
+        version: RowVersion<T>,
+    ): RowVersion<T>? {
+        val draft = fetch(version)
+        val designRowId = draft.contextData.designRowId
+        return if (branch.designId == null && draft.contextData.officialRowId == null && designRowId is IntId) {
+            queryDesignRowVersion(designRowId)
+        } else if (draft.isDesign && draft.contextData.officialRowId != null) {
+            if (designRowId is IntId) queryDesignRowVersion(designRowId) else null
+        }
+        else fetchVersion(branch.official, version.id)
+    }
+
+    private fun queryDesignRowVersion(designRowId: IntId<T>): RowVersion<T> = jdbcTemplate.queryOne(
+        "select id, version from ${table.fullName} where id = :designRowId", mapOf("designRowId" to designRowId.intValue)
+    ) { rs, _ ->
+        rs.getRowVersion("id", "version")
+    }
+
```

Comment:
En ole ihan satavarma, onko tämän funktion oikein olla olemassa lainkaan. Se tekee suunnilleen samat asiat mutta vähän eri kautta kuin asDesignDraft() jne. -funktiot, ts. hakee sen version, jonka katsotaan olevan tätä draftia vastaava virallinen versio, ja siis se, jota pitää muuttaa, jotta ID:t pysyy kuosissa. Tapaukset ylhäältä alaspäin katsoen siis:
- design-haarassa luodun olion julkaisu mainissa
- mainissa olemassaolleen olion julkaisu designissa ensimmäistä kertaa tai myöhempiä kertoja
- muut, ts. haarojen välillä siirtymättömien olioiden julkaisut

Reply:
Hmm.. tätä ei minusta pitäisi tarvita mihinkään, koska jokainen olio kantaa jo mukanaan contextdatan ja se sisältää tämän tiedon. Eli jos haetaan vaikkapa design-draft olio, sen contextDatasta löytyy designRowId joka viittaa tuohon design-riviin ja officialRowId joka viittaa viralliseen riviin.

Reply:
Mietin tätä vähän tarkemmin. Periaatteessa sen pitäisi mennä näin:

Kun ollaan julkaisemassa, meillä on joko kyseessä jonkun designin draft tai mainin draft. Ainoa paikka mihin sitä voidaan julkaista on ko. branchin official. Kaikki muut siirtymät tehdään draftaus-vaiheessa, eli luomalla ensin uusi draft oikeaan branchiin. Eli oikeastaan kaikissa tilanteissa, tuon draft-olion branch on oikea branch ja nykyisin julkaisufunkkarille passattava branch täytyy olla sama kuin validaatioversiolla haetun olion palauttama branch. Tälle voisi joko lisätä assertin sanity checkinä, tai ehkä vain poistaa julkaisu-metodista argumentin, koska branch pitäisi olla olion mukainen.

Kun julkaisu ylikirjoittaa olemassaolevan official-rivin, tämä tapahtuu päivittämällä kontekstin tiedot ja sanomalla DAO:on update. Tässä kohtaa on eri caset:

1. Jos draft-riviä vastaavaa officialia (kyseisessä branchissa) ei ole, ID säilyy mutta draft-flagi asetetaan falseksi. Viitettä branchin omaan official-riviin ei voi tässä kohtaa olla koska sellaista ei ole, mutta designilla voi olla viite main-puolen viralliseen ja tuon tiedon pitäisi säilyä kenttänä entisellään. Operaation jälkeen, mitään riviä ei tarvitse poistaa, koska draft vain muuttui officialiksi.
2. Jos draft-riviä vaastava official on olemassa, sen ID asetetaan draft-olion kontekstiin (ja draft-flag taas falseksi), jolloin update itse asiassa kohdistuu official-riviin. Muilta osin homma menee kuin yllä. Operaation jälkeen, itse draft-rivi kuitenkin kummitelee edelleen joten se pitää poistaa. Tämän tunnistuksessa taitaa nykyisellään olla bugi noiden designien kohdalla koska se oletti voivansa tarkastaa tilanteen katsomalla official vs draft id:tä, mutta tuon voi yhtä hyvin katsoa siitä vaihtuiko draftin id kontekstin päivityksessä. Lisäksi, jos kyseessä oli main-draft rivi joka on tullut designista, pitää myös design-rivi poistaa. Tuonkin voi iffittää kontekstin tiedoilla.

Rempattu julkaisulogiikkaa tähän malliin: https://github.com/finnishtransportagency/geoviite/pull/1289/files#r1629796923